src/index.spec.js

```diff
@@ -1320,68 +1331,77 @@ describe('DataModel', () => {
+                { profit: -21, sales: 25, first: 'White', second: 'walls' },
                 { profit: 21, sales: 25, first: 'White', second: 'walls' },
-                { profit: 21, sales: 25, first: 'White', second: 'walls' }
+                { profit: -14, sales: 25, first: 'White', second: 'walls' }
             ];
             const schema1 = [
                 { name: 'profit', type: 'measure' },
                 { name: 'sales', type: 'measure' },
                 { name: 'first', type: 'dimension' },
                 { name: 'second', type: 'dimension' },
             ];
-            const dataModel = new DataModel(data1, schema1, { name: 'Yo' });
-            const bin = dataModel.bin('profit', { binSize: 5, name: 'sumField' });
-            let fieldData = bin.getFieldspace().fields.find(field => field.name() === 'sumField').partialField.data;
-            let expData = ['10-15', '15-20', '15-20', '15-20', '10-15', '15-20', '20-25', '15-20', '20-25', '20-25'];
-            expect(expData).to.deep.equal(fieldData);
-            expect(bin.getFieldspace().fields.find(field => field.name() === 'sumField').bins().mid)
-                            .to.deep.equal([12.5, 17.5, 22.5]);
-            expect(bin.getFieldspace().fields.find(field => field.name() === 'sumField').bins().range)
-                            .to.deep.equal([10, 15, 20, 25]);
+            const dm = new DataModel(data1, schema1, { name: 'Yo' });
+
+            let binnedDm = dm.bin('profit', { buckets: [0, 6, 12, 20], name: 'sumField' });
+            let newField = binnedDm.getFieldspace().fields.find(field => field.name() === 'sumField');
+            let expData = ['6-12', '12-20', '12-20', '12-20', '6-12', '-21-0', '20-22', '12-20', '-21-0', '20-22',
+                '-21-0'];
+            let expBins = [-21, 0, 6, 12, 20, 22];
+            expect(newField.data()).to.deep.equal(expData);
+            expect(newField.bins()).to.deep.equal(expBins);
+
+            binnedDm = dm.bin('profit', { binsCount: 5, start: 1, name: 'sumField1' });
+            newField = binnedDm.getFieldspace().fields.find(field => field.name() === 'sumField1');
+            expData = ['6-15', '15-24', '15-24', '15-24', '6-15', '-21--12', '15-24', '15-24', '-21--12', '15-24',
+                '-21--12'];
+            expBins = [-21, -12, -3, 6, 15, 24];
+            expect(newField.data()).to.deep.equal(expData);
+            expect(newField.bins()).to.deep.equal(expBins);
+
+            binnedDm = dm.bin('profit', { binSize: 10, start: -1, end: 30, name: 'sumField2' });
+            newField = binnedDm.getFieldspace().fields.find(field => field.name() === 'sumField2');
+            expData = ['9-19', '9-19', '9-19', '9-19', '9-19', '-21--11', '19-29', '9-19', '-21--11', '19-29',
+                '-21--11'];
+            expBins = [-21, -11, -1, 9, 19, 29, 39];
+            expect(newField.data()).to.deep.equal(expData);
+            expect(newField.bins()).to.deep.equal(expBins);
```

Comment:
Need a test case where data value and end / start of a bucket has the same value.